crates/sui-replay-2/src/data_store.rs

```diff
@@ -0,0 +1,428 @@
+                    })?;
+            let (page_info, data) = pkg_versions.into_parts();
+            for (pkg, pkg_version, epoch) in data {
+                let package_obj = pkg.ok_or_else(|| ReplayError::PackageNotFound {
+                    pkg: pkg_address.to_string(),
+                })?;
+                let package_obj = Object::try_from(package_obj).map_err(|e| {
+                    ReplayError::ObjectConversionError {
+                        id: pkg_address.to_string(),
+                        err: format!("{:?}", e),
+                    }
+                })?;
+
+                trace!(
+                    "Collecting system package: {}[{} - {}], {:?}",
+                    package_obj.id(),
+                    package_obj.version(),
+                    pkg_version,
+                    epoch,
+                );
+                // TODO: fix this. If epoch is missing come up with something....
+                let epoch = epoch.unwrap_or(0);
+                // epoch.ok_or_else(|| ReplayError::MissingPackageEpoch {
+                //     pkg: pkg_address.to_string(),
+                // })? as u64;
+                packages.push((package_obj, epoch));
+            }
+            if page_info.has_next_page {
+                pagination = PaginationFilter {
+                    direction: Direction::Forward,
+                    cursor: page_info.end_cursor.clone(),
+                    limit: None,
+                };
+            } else {
+                break;
+            }
+        }
+        debug!("End load system package");
+        Ok(packages)
+    }
+
+    /// Load all versions of all system packages
+    pub async fn load_system_packages(
+        &self,
+    ) -> Result<BTreeMap<ObjectID, BTreeMap<u64, Object>>, ReplayError> {
+        debug!("Start load system packages");
+        let mut system_packages = BTreeMap::new();
+        for pkg_id in sui_framework::BuiltInFramework::all_package_ids() {
+            let packages = self.load_system_package(pkg_id).await?;
+            let all_versions = packages
+                .into_iter()
+                .map(|(pkg, epoch)| {
+                    trace!("{}[{}] - {}", pkg.id(), pkg.version(), epoch);
+                    (epoch, pkg)
+                })
+                .collect();
+            system_packages.insert(pkg_id, all_versions);
+        }
+        debug!("End load system packages");
+        Ok(system_packages)
+    }
+
+    // Load a set of (ObjectId, version) objects
+    pub async fn load_objects(
+        &self,
+        object_inputs: &BTreeSet<InputObject>,
+    ) -> Result<Vec<(ObjectID, Option<u64>, Object)>, ReplayError> {
+        debug!("Start load objects");
+        let mut objects = vec![];
+        for object_input in object_inputs {
+            let InputObject {
+                object_id, version, ..
+            } = object_input;
+            let address = Address::new((*object_id).into_bytes());
```

Comment:
(not a blocker for this PR)

Does `Address` not have a `From` instance for `ObjectID`? If not may be worth adding one...

Reply:
not that I know, sorry. The thing here is aggravated by the fact that `Address` is an SDK type and `ObjectID` a sui type. So not all conversion from/to are supported.